### PR TITLE
CMake: sanitise Visual Studio's compile_commands.json for use by Clang

### PR DESCRIPTION
**Pull request by PhilipOakley**
The Visual Studio command line saved by commit 409047a2b3 (cmake: create
compile_commands.json by default, 2021-06-06) is not compatible with
tools that use Clang for their code analysis.

Provide a shell script that will sanitise the compile_commands.json file.

Signed-off-by: Philip Oakley <philipoakley@iee.email>

---
This is should be useful to potential Git-for-Windows contributors who wish to
browse the code with the open source Sourcetrail interactive dependency graph
viewer https://www.sourcetrail.com/

If there are no issues I'll also send it upstream (GitGitGadget or Patches)

### CODE REVIEW

**Comment by PhilipOakley** [comment]
This is in parallel with #3306

**Comment by dscho** [comment]
Thank you for working on this. However, I am not sure that this attacks the correct layer. It seems as if you let CMake generate the files targeting the MS Visual C compiler, but then expect to use the files as if clang had been targeted.

So I wonder whether a better idea would be to follow [Visual Studio's official guidance targeting Clang/LLVM](https://docs.microsoft.com/en-us/cpp/build/clang-support-cmake) instead? The way I read this documentation, the result should be `compile_commands.json` files that SourceTrail can handle, without us having to hard-code options that might, or might not, change in the future, totally out of our control.

**Comment by PhilipOakley** [comment]
> Thank you for working on this. However, I am not sure that this attacks the correct layer. It seems as if you let CMake generate the files targeting the MS Visual C compiler, but then expect to use the files as if clang had been targeted.
> 
> So I wonder whether a better idea would be to follow [Visual Studio's official guidance targeting Clang/LLVM](https://docs.microsoft.com/en-us/cpp/build/clang-support-cmake) instead? The way I read this documentation, the result should be `compile_commands.json` files that SourceTrail can handle, without us having to hard-code options that might, or might not, change in the future, totally out of our control.

I think we are viewing this from different directions. For me, Sourcetrail is being used as a visualiser of the code we have, and I'm, in particular, trying to ensure we have a range of smooth entry slipways for potential contributors and lurkers to explore the code (using one of the VARK approaches..). 

You are right that there is a mismatch between regular Visual Studio tools and its compilation, and Clang based tools, such as Sourcetrail. Sourcetrail's starting point is usually the compile_commands.json file. One other option is that Sourcetrail has also provided a Visual Studio Extension that will somehow extract what it needs from the .sln (and proj files) for the main program to run, but the extension also requires extra user install actions that can confuse (it confused me). 

Meanwhile, our CMake doesn't generate an obvious .sln file for Sourcetrail. Hence this script is provided as a simple sanitiser as one method to get folks started. I don't think we should be changing our CMake to retarget Clang when it's purpose was to target core VS anyway. 

> options that might, or might not, change in the future
We have always been hanging by that shoogly peg that we are at the mercy of the whims of Microsoft (e.g. see the change to Ninja). 

As a side question, Have you used any other code visualisers that may provide an alternative? 

[I'm away for a log weekend in a few hours]

**Comment by dscho** [comment]
> our CMake doesn't generate an obvious .sln file for Sourcetrail.

I think this is the thing to concentrate on. Since our CMake configuration (which is not even obvious to discover in the first place!) does not generate `.sln` files, I think here is a grand opportunity for a step-by-step guide to setting things up for Sourcetrail, probably as a Wiki page. And this documentation should then also include the guidance how to set up CMake to generate files targeting Clang/LLVM.

In other words, I still think that there are better ways than what this PR tries to do.

**Comment by PhilipOakley** [comment]
The SourceTrail team recently [announced ](https://www.sourcetrail.com/blog/discontinue_sourcetrail/)that they would no longer be continuing the development of the SourceTrail tool, which is 'unfortunate', so I'll close this PR at it's current state. 

It may come in useful as a starting point if others need to sanitise the VS output of the `compile_commands.json` files to match Clang's expectations.
